src/main/java/com/artipie/asto/Simple.java

```diff
@@ -0,0 +1,157 @@
+import hu.akarnokd.rxjava3.jdk8interop.SingleInterop;
+import io.reactivex.rxjava3.core.Completable;
+import io.reactivex.rxjava3.core.Flowable;
+import io.reactivex.rxjava3.core.Single;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Flow;
+import java.util.stream.Collectors;
+import org.reactivestreams.FlowAdapters;
+
+/**
+ * Simple storage, in files.
+ *
+ * @since 0.1
+ */
+public final class Simple implements Storage {
+
+    /**
+     * Where we keep the data.
+     */
+    private final Path dir;
+
+    /**
+     * Ctor.
+     *
+     * @throws IOException If fails
+     */
+    public Simple() throws IOException {
+        this(Files.createTempDirectory("asto"));
+    }
+
+    /**
+     * Ctor.
+     *
+     * @param path The path to the dir
+     */
+    public Simple(final Path path) {
+        this.dir = path;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(final String key) {
+        return (CompletableFuture<Boolean>) Single.fromCallable(
+            () -> {
+                final Path path = Paths.get(this.dir.toString(), key);
+                return Files.exists(path);
+            }).to(SingleInterop.get());
+    }
+
+    @Override
+    public CompletableFuture<Collection<String>> list(final String prefix) {
+        return (CompletableFuture<Collection<String>>) Single.fromCallable(
+            () -> {
+                if (!prefix.endsWith("/")) {
+                    throw new IllegalArgumentException(
+                        String.format(
+                            "The prefix must end with a slash: \"%s\"",
+                            prefix
+                        )
+                    );
+                }
+                final Path path = Paths.get(this.dir.toString(), prefix);
+                final Collection<String> keys = Files.walk(path)
+                    .filter(Files::isRegularFile)
+                    .map(Path::toString)
+                    .map(
+                        p -> p.substring(
+                            path.toString().length() - prefix.length() + 1
+                        )
+                    )
+                    .collect(Collectors.toList());
+                Logger.info(
+                    this,
+                    "Found %d objects by the prefix \"%s\" in %s by %s: %s",
+                    keys.size(), prefix, this.dir, path, keys
+                );
+                return keys;
+            }).to(SingleInterop.get());
+    }
+
+    @Override
+    public CompletableFuture<Void> save(final String key, final Flow.Publisher<Byte> content) {
+        final Completable result = Flowable.fromPublisher(FlowAdapters.toPublisher(content))
+            .toList()
+            .map(bytes -> bytes.toArray(new Byte[0]))
```

Comment:
@Sammers21 these two maps can be converted into single one

Reply:
Yes, but should they?

Reply:
@Sammers21 I think yes in this case, since both maps exists here for one purpose only: convert list of bytes `ByteArray` class.

Reply:
Agree